src/entry.c

```diff
@@ -0,0 +1,356 @@
+
+    bool update_value = value ? true : false;
+    long long expiration_time = entryGetExpiry(e);
+    bool update_expiry = (expiry != expiration_time) ? true : false;
+    if (!update_value && !update_expiry)
+        return e;
+    expiration_time = expiry;
+    value = update_value ? value : entryGetValue(e);
+    size_t expiry_size = (expiration_time != EXPIRY_NONE) ? sizeof(expiration_time) : 0;
+    int field_sds_type = sdsReqType(sdslen(field));
+    if (field_sds_type == SDS_TYPE_5 && (expiry_size > 0)) {
+        field_sds_type = SDS_TYPE_8;
+    }
+    size_t field_size = sdsHdrSize(field_sds_type) + sdsalloc(field) + 1;
+    size_t value_len = value ? sdslen(value) : 0;
+    size_t embedded_value_size = value ? sdsReqSize(value_len, SDS_TYPE_8) : 0;
+
+    size_t required_embedded_size = field_size + embedded_value_size + expiry_size;
+    size_t current_embedded_allocation_size = entryHasValuePtr(e) ? 0 : entryMemUsage(e);
+    /* // We will create a new entry in the following cases:
+     * 1. In the case were we add or remove expiration.
+     * 2. in the case were we are NOT migrating from an embedded entry to an embedded entry with ~the same size. */
+    bool create_new_entry = (update_expiry && (entryGetExpiry(e) == EXPIRY_NONE || expiration_time == EXPIRY_NONE)) ||
+                            !(update_value && !entryHasValuePtr(e) &&
+                              required_embedded_size <= EMBED_VALUE_MAX_ALLOC_SIZE &&
+                              required_embedded_size <= current_embedded_allocation_size &&
+                              required_embedded_size >= current_embedded_allocation_size * 3 / 4);
+
+    if (!create_new_entry) {
+        /* In this case we are sure we do not have to allocate new entry, so expiry must already be set. */
+        if (update_expiry) {
+            serverAssert(entryHasExpiry(e));
+            char *buf = sdsAllocPtr(e);
+            if (entryHasValuePtr(e)) buf -= sizeof(sds);
+            buf -= sizeof(expiry);
+            memcpy(buf, &expiry, sizeof(expiry));
+        }
+        /* In this case we are sure we do not have to allocate new entry, so value must already be set or we have enough room to embed it. */
+        if (update_value) {
+            if (entryHasValuePtr(e)) {
+                sds *value_ref = entryGetValueRef(e);
+                sdsfree(*value_ref);
+                *value_ref = value;
+            } else {
+                /* Skip field content, field null terminator and value sds8 hdr. */
+                sds old_value = entryGetValue(e);
+                /* We are using the same entry memory in order to store a potentially new value.
+                 * In such cases the old value alloc was adjusted to the real buffer size part it was embedded to.
+                 * since we can potentially write here a smaller value, which requires less allocation space, we would like to
+                 * inherit the old value memory allocation size. */
```

Comment:
I can't tell what's happening here. I'm guessing that the second line should be the start of a sentence, but it's possible it's the continuation for `embedded to`...
```suggestion
                 * In such cases the old value alloc was adjusted to the real buffer size part it was embedded to.
                 * Since we can potentially write here a smaller value, which requires less allocation space, we would like to
                 * inherit the old value memory allocation size. */
```